console-subscriber/src/aggregator/mod.rs

```diff
@@ -274,24 +282,48 @@ impl Aggregator {
     /// Add the task subscription to the watchers after sending the first update
     fn add_instrument_subscription(&mut self, subscription: Watch<proto::instrument::Update>) {
         tracing::debug!("new instrument subscription");
-
-        let task_update = Some(self.task_update(Include::All));
-        let resource_update = Some(self.resource_update(Include::All));
-        let async_op_update = Some(self.async_op_update(Include::All));
         let now = Instant::now();
 
-        let update = &proto::instrument::Update {
-            task_update,
-            resource_update,
-            async_op_update,
-            now: Some(self.base_time.to_timestamp(now)),
-            new_metadata: Some(proto::RegisterMetadata {
-                metadata: (*self.all_metadata).clone(),
-            }),
+        let update = loop {
+            let update = proto::instrument::Update {
+                task_update: Some(self.task_update(Include::All)),
+                resource_update: Some(self.resource_update(Include::All)),
+                async_op_update: Some(self.async_op_update(Include::All)),
+                now: Some(self.base_time.to_timestamp(now)),
+                new_metadata: Some(proto::RegisterMetadata {
+                    metadata: (*self.all_metadata).clone(),
+                }),
+            };
+            let el = update.encoded_len();
+            if el < MAX_MESSAGE_SIZE {
+                // normal case
+                break Some(update);
+            }
+            // If the grpc message is bigger than tokio-console will accept throw away the oldest
+            // inactive data and try again
+            self.retention /= 2;
+            self.cleanup_closed();
+            tracing::debug!(
+                retention = ?self.retention,
+                message_size = el,
+                max_message_size = MAX_MESSAGE_SIZE,
+                "Message too big, reduced retention",
+            );
+
+            if self.retention <= MIN_RETENTION {
+                self.retention = MIN_RETENTION;
+                break None;
+            }
```

Comment:
I don't think hard-coding the minimum retention to 1 second makes sense. Updates are only published to watchers every time the `Aggregator`'s `publish_interval` elapses, which is a user-configurable value that defaults to 2 seconds. If we publish updates every two seconds, but only retain data for one second, we'll be throwing away half of the data about events that occurred since the last update, which seems like unacceptable data loss to me.

Instead, I think the minimum allowable retention duration probably ought to be `self.publish_interval`. 
```suggestion
            if self.retention <= self.publish_interval {
                self.retention = self.publish_interval;
                break None;
            }
```

This way, we will always retain data for at least long enough to include it in an update --- we would be discarding historical data that future clients might be interested in, but we would at least always be ensuring that actively connected clients are still receiving all real-time data as it's produced. 

Incidentally, this would mean that users could also tune the publish interval to allow running with lower retention values.